packages/integration-sdk-runtime/src/execution/duplicateKeyTracker.ts

```diff
@@ -132,8 +132,98 @@ export type DuplicateEntityReport = {
   _key: string;
   rawDataMatch: boolean;
   propertiesMatch: boolean;
+  rawDataDiff?: string;
+  propertiesDiff?: string;
 };
 
+type DiffType =
+  | 'missing_in_src'
+  | 'missing_in_dest'
+  | 'type_mismatch'
+  | 'value_mismatch';
+
+interface ObjectDiff {
+  [key: string]: {
+    type: DiffType;
+    valueTypes?: { src: string; dest: string };
+  };
+}
+
+/**
+ * Compares two objects and returns the differences between them.
+ *
+ * @param {unknown} src - The source object to compare.
+ * @param {unknown} dest - The destination object to compare.
+ * @param {string} [path=''] - The base path for keys, used for tracking nested object differences.
+ * @returns {ObjectDiff} An object representing the differences between `src` and `dest`.
+ *   Each key corresponds to a path in the objects, with details about the type of difference.
+ *
+ * @example
+ * const src = { a: 1, b: { c: 2 } };
+ * const dest = { a: 1, b: { c: 3 }, d: 4 };
+ * const result = diffObjects(src, dest);
+ * console.log(result);
+ * // Output:
+ * // {
+ * //   "b.c": { type: "value_mismatch" },
+ * //   "d": { type: "missing_in_src" }
+ * // }
+ */
+export function diffObjects(
+  src: unknown,
+  dest: unknown,
+  path: string = '',
+): ObjectDiff {
+  const diff = {};
+
+  // Helper to add differences
+  const addDiff = (
+    key: string,
+    type: DiffType,
+    valueTypes?: { src: string; dest: string },
+  ) => {
+    diff[key] = { type, valueTypes };
+  };
+
+  // Iterate through the keys of both objects
+  const allKeys = new Set([
+    ...Object.keys(src || {}),
+    ...Object.keys(dest || {}),
+  ]);
+
+  const isObject = (val: unknown): val is Record<string, unknown> =>
+    typeof val === 'object' && val !== null;
+
+  for (const key of allKeys) {
+    const fullPath = path ? `${path}.${key}` : key;
+    const valSrc = src?.[key];
+    const valDest = dest?.[key];
+
+    if (valSrc === undefined) {
+      addDiff(fullPath, 'missing_in_src');
+    } else if (valDest === undefined) {
+      addDiff(fullPath, 'missing_in_dest');
+    } else if (typeof valSrc !== typeof valDest) {
+      addDiff(fullPath, 'type_mismatch', {
+        src: typeof valSrc,
+        dest: typeof valDest,
+      });
+    } else if (Array.isArray(valSrc) && Array.isArray(valDest)) {
+      if (JSON.stringify(valSrc) !== JSON.stringify(valDest)) {
+        addDiff(fullPath, 'value_mismatch');
+      }
+    } else if (isObject(valSrc) && isObject(valDest)) {
+      // Recursive comparison for nested objects
+      const nestedDiff = diffObjects(valSrc, valDest, fullPath);
```

Comment:
How can we prevent infinite recursion if a property in an object references itself?

Reply:
Hmm, I don't think that's possible, in raw data certainly not because it's parsed from a string (the response) so each nested object is a unique object. And in the entity properties we shouldn't have that issue either because we only accept string, number and boolean (and array of those).
We could prevent it by keeping track of every object we encounter but that adds a layer of complexity because we'll be using more memory